File: python/unionfind.py
"""
NAME
    union_find - union-find data structure 

DESCRIPTION 
    This module implements union-find data structure including 
        1) find with path compression 
        2) union with rank. 

    * UnionFind      union-find via array implementation
    * UnionFindDict  union-find via dictionary implementation

CLASSES
    class UnionFind(n)
     |  Return a union-find data structure 
     |  
     |  Methods defined here: 
     | 
     |  find(p)
     |      find the parent of p
     | 
     |  union(p, q) 
     |      connect p and q into one component 

    class UnionFindDict()
     |  Return a union-find data structure 
     |  
     |  Methods defined here: 
     | 
     |  find(p) 
     |      find the parent of p 
     | 
     |  union(p, q)  
     |      connect p and q into one component
"""

class UnionFind:
    """Union-Find via array implementation"""

    def __init__(self, n: int):
        self.parent = list(range(n))
        self.rank = [1] * n

    def find(self, p: int) -> int:
        """Find with path compression"""
        if p != self.parent[p]:
            self.parent[p] = self.find(self.parent[p])
        return self.parent[p]

    def union(self, p: int, q: int) -> bool:
        """Union with rank"""
        prt, qrt = self.find(p), self.find(q)
        if prt == qrt: return False
        if self.rank[prt] > self.rank[qrt]: prt, qrt = qrt, prt 
        self.parent[prt] = qrt
        self.rank[qrt] += self.rank[prt]
        return True


class UnionFindDict:
    """Union-Find via dictionary implementation"""

    def __init__(self): 
        self.parent = {}
        self.rank = defaultdict(lambda: 1)
    
    def find(self, p):
        """Find with path compression"""
        if p not in self.parent: self.parent[p] = p
        elif p != self.parent[p]:
            self.parent[p] = self.find(self.parent[p])
        return self.parent[p]
    
    def union(self, p, q):
        """Union with rank"""
        prt, qrt = self.find(p), self.find(q)
        if prt == qrt: return False 
        if self.rank[prt] > self.rank[qrt]: prt, qrt = qrt, prt 
        self.parent[prt] = qrt
        self.rank[qrt] += self.rank[prt]
        return True 

File: cpp/unionfind.cpp
#include <algorithm>
#include <numeric>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

class UnionFind {
    /*Union-Find implemented via array*/
public: 
    vector<int> parent, rank; 
    UnionFind(int n) {
        parent.resize(n); 
        iota(begin(parent), end(parent), 0); 
        rank.resize(n); 
        fill(rank.begin(), rank.end(), 1); 
    } 
    
    int find(int p) {
        /* find with path compression */
        if (parent[p] != p) 
            parent[p] = find(parent[p]); 
        return parent[p]; 
    }
    
    bool connect(int p, int q) {
        /* union with rank */
        int prt = find(p), qrt = find(q); 
        if (prt == qrt) return false; 
        if (rank[prt] > rank[qrt]) swap(prt, qrt);
        parent[prt] = qrt; 
        rank[qrt] += rank[prt]; 
        return true; 
    }
};


class UnionFindDict {
    /*Union-Find implemented via dictionary*/
public: 
    unordered_map<string, string> parent; 
    unordered_map<string, int> rank; 
    
    string find(string p) {
        if (!parent.count(p)) parent[p] = p, rank[p] = 1; 
        else if (p != parent[p]) 
            parent[p] = find(parent[p]); 
        return parent[p]; 
    }
    
    bool connect(string p, string q) {
        string prt = find(p), qrt = find(q); 
        if (prt == qrt) return false; 
        if (rank[prt] > rank[qrt]) swap(prt, qrt); 
        parent[prt] = qrt; 
        rank[qrt] += rank[prt]; 
        return true; 
    }
}; 
